### DISCUSSION TITLE
How to use config option `fullTracer.printOpcodes` in `zkevm-proverjs`

### DISCUSSION THREAD

**Discussion by mooori** [0xPolygonHermez/pilcom#62] [Q&A] [answered]
For benchmarking of NEAR’s ZK WASM I’m trying to make `zkevm-proverjs` print opcodes using [`fulltracer.printOpcodes`](https://github.com/0xPolygonHermez/zkevm-proverjs#verbose-configuration). [This script](https://github.com/near/wasmtime/blob/e7eb24df17770ae3c12b5a91ba382069806d57ca/tests/zkasm/run-tests-zkasm.js#L114) triggers execution and I’m using the following config:

```js
const config = {
    debug: true,
    debugInfo: {
        inputName: path.basename(pathTest),
    },
    tracer: true,
    stepsN: 8388608,
    assertOutputs: false,
    verboseOptions: {
        fulltracer: {
            enable: true,
            enableOpcodes: true,
            printOpcodes: true,
            bytecode: true,
            outputs: true,
            saveOutputs: true
        }
    }
};
```

### Tracer objects are constructed but `onOpcode` isn’t called

By manual debugging I could check that both [`FullTracer`](https://github.com/0xPolygonHermez/zkevm-proverjs/blob/97927e70761956a1091e1b5d2a93c48134fb62be/src/sm/sm_main/debug/full-tracer.js#L44) and [`VerboseTracer`](https://github.com/0xPolygonHermez/zkevm-proverjs/blob/97927e70761956a1091e1b5d2a93c48134fb62be/src/sm/sm_main/debug/verbose-tracer.js#L18) are constructed. However [`onOpcode`](https://github.com/0xPolygonHermez/zkevm-proverjs/blob/97927e70761956a1091e1b5d2a93c48134fb62be/src/sm/sm_main/debug/full-tracer.js#L540) (which should print the opcode) is not called.

### Possible issues

- Some other steps are needed to emit `onOpcode` events?
- Is `printOpcodes` used in a wrong location?
- Is there other configuration that needs to be enabled to make `printOpcodes` work?
- Are logs written but I just cannot find them?
  - I would assume that’s rather not the problem as I inserted logs into `zkevm-proverjs` to print the directory where logs should be written. After execution these directories are empty.

### Smoke test: `stats: true`

This works for me. When adding `stats: true` to above config, data is written to `node_modules/0xpolygonhermez/zkevm-proverjs/src/sm/sm_main/logs-stats/*.json`.

### Branch `feature/64bits`

Note that `near/wasmtime` uses branch [`feature/64bits`](https://github.com/near/wasmtime/blob/e7eb24df17770ae3c12b5a91ba382069806d57ca/tests/zkasm/package.json#L17) of `zkevm-proverjs`.

**Comment by krlosMata** [comment, marked as answer]
hi !!
## Brief explanation of modules:
- `fullTracer`: Class to get execution traces of the batch execution. Heavily focus to get EVM traces like `geth` does for Ethereum. It is like an EVM trace extended wverboseOptionsith some zkEVM parameters
- `verboseOptions`: property of the `config`. Useful to print data on the terminal (or save data) that is collected by the `fullTracer`. Very useful to debug without doing breakpoints.

## Enable fullTracer
- `main_executor.js` should be run with the `-T` flag: https://github.com/0xPolygonHermez/zkevm-proverjs/blob/feature/64bits/src/main_executor.js#L35
  - note that `fullTracer` has not been ported to 64 bits registers
- `fullTracer` works by triggering its events (`eventLog`) from the `.zkasm` code: https://github.com/0xPolygonHermez/zkevm-rom/blob/main/main/process-tx.zkasm#L381
  - this basically means that every time this line of the `.zkasm` is triggered --> the `sm_main` calls the `fullTracer` to that specific function --> https://github.com/0xPolygonHermez/zkevm-proverjs/blob/feature/64bits/src/sm/sm_main/sm_main_exec.js#L1753
    - then, inside the `fullTracert` --> the `onOpcode` is called
    - inside the `onOpcode` all the necessary data is collected to create the EVM execution trace opcode step
      - note that all the data collected depends heavily on the `.zkasm` variables defined     

## Suggestion
- take the `fullTracer` class and add a new function like `onOpcode64`
- in the `.zkasm` add the necessary event `$${eventLog(onOpcode64())}`
- collect necessary data in the `fullTracer` function    

hope it helps xD
